### ISSUE TITLE
Link widget - Blocks cannot be updated when memcached is enabled Prestashop 1.7

### ISSUE THREAD

**Issue by ets-soft** [PrestaShop/PrestaShop#18171]
#### Describe the bug

I enabled caching with CacheMemcached, my memcached server has been correctly configured and working well. But since the caching is enabled, sometimes when I save products, disable/enable module... it doesn't take effect! There is no change on the FO when I update product information (sometimes), I disable a module and reload the BO the module keep enabled...
I have checked some core files of Prestashop, I think the problem is in doctrine, when an SQL is executed via doctrine it doesn't clear all related SQL cached queries!  

#### Expected behavior

All related cached SQL queries must be cleared when an Insert, update, delete, etc. queries are executed

#### Steps to Reproduce
>Edit by Khouloud

the cache issue is happening on another area such as module link list this: https://prnt.sc/rjjfs3
Can you follow:
1. enable your Memcached 
2. Design > Link widget
3. Try to edit some  link (title, URL...) then save
4. Go to the Front office and review the links. You will notice there are no changes in FO
---------------------------

1. Go to "Advanced parameters / Performance / Caching"
2. Set "Use cache" to "Yes"
3. Select caching system as "CacheMemcached"
4. Configure your memcached server
5. After you have successfully enabled memcached, try to disable a certain module via module list in BO then reload the BO you will see the module keep enabled! Try to update product information via BO and review the product on FO you will see product info is not updated (sometimes)
 
#### Additional information

* PrestaShop version: 1.7
* PHP version: 7.0

**Comment by khouloudbelguith** [comment]
Hi @ets-soft,

Thanks for your report.
I did not manage to reproduce the issue with PS1.7.6.4.
I attached a screen record
[Untitled_ Mar 17, 2020 10_44 AM.webm.zip](https://github.com/PrestaShop/PrestaShop/files/4342662/Untitled_.Mar.17.2020.10_44.AM.webm.zip)
Could you please provide us with more info? We need more details to understand how we can reproduce your issue:
 - host
 - server setup and configuration
 - PrestaShop version (source)
 - debug mode report
 - PHP error logs
 - apache error log
 - javascript console log
 - screenshots

Don't you know how to get this information? Please read the following article:
http://build.prestashop.com/howtos/misc/how-to-create-bug-report/

Thanks!

**Comment by ets-soft** [comment]
hi @khouloudbelguith 
Thanks for checking, I just review the problem, seem it's ok when updating product info but the cache issue is happening on other area such as module link list this: https://prnt.sc/rjjfs3
Can you follow:
1. enable your memcached 
2. Design > Link widget
3. Try to edit some  link (title, URL...) then save
4. Go to Front office and review the links. You will notice there is no changes in FO

Thanks

**Comment by khouloudbelguith** [comment]
@ets-soft, thanks for these clarifications.
I manage to reproduce the issue with PS1.7.5.2 and PS1.7.6.4.
[Untitled_ Mar 23 2020 2_17 PM.webm.zip](https://github.com/PrestaShop/PrestaShop/files/4369586/Untitled_.Mar.23.2020.2_17.PM.webm.zip)

I’ll add this to the debug roadmap so that it’s fixed. If you have already fixed it on your end or if you think you can do it, please do send us a pull request!
Thanks!

**Comment by ets-soft** [comment]
hi @khouloudbelguith 
I think the problem may not only happen in the link widget title, it's only 1 example case. Please tell the dev team to check the whole caching process deeply to make sure it works correctly everywhere.
Actually, I tried to look into the code of Prestashop but could not figure out how to fix the issue so I'm looking forward to a fix from Prestashop as this is really important feature and I really need it for my customers' website.

Thanks

**Comment by khouloudbelguith** [comment]
@ets-soft, thanks for these clarifications.

**Comment by LukasK13** [comment]
I can confirm this issue on PS 1.7.7.0

**Comment by Renrhaf** [comment]
Confirming too on PS 1.7.7.4

**Comment by ets-soft** [comment]
> Confirming too on PS 1.7.7.4

Memcached is greatly necessary to improve website speed as well as to save MySQL resources so we can serve more users at the same time, however it's not stable on PrestaShop, once enable Memcached we got a lot of problems so we had to disabled it completely. Hope PrestaShop development team focus on Memcached and make it usable soon!

**Comment by Renrhaf** [comment]
Yeah I'm pretty surprised that this issue is classified as minor.
I had configured Memcached properly, but now I can't use it anymore due to this bug and some instabilities I have been facing.

**Comment by kpodemski** [comment]
There are a lot of issues with Memcache, not only with `ps_linklist` but with many other third party modules too, I'm pretty sure this is Minor because Memcache is not popular in PrestaShop ecosystem, most of the time when I see Memcache enabled, it is in store with 100 orders / month and almost clean installation ;)

**Comment by Renrhaf** [comment]
@kpodemski thanks for the tip, do you suggest using any other caching layer than Memcached that would have better support in Prestashop ?

**Comment by kpodemski** [comment]
@Renrhaf what I found most performant was to use some full page cache module and disable smarty cache, checkout page might need additional changes but overall I highly recommend full page caching + maybe CloudFlare for assets, of course this is just for start, there is a lot what we can do to improve performance :)

**Comment by ets-soft** [comment]
> There are a lot of issues with Memcache, not only with `ps_linklist` but with many other third party modules too, I'm pretty sure this is Minor because Memcache is not popular in PrestaShop ecosystem, most of the time when I see Memcache enabled, it is in store with 100 orders / month and almost clean installation ;)

@kpodemski I agree Memcached is not popular in PrestaShop as there are not so many users enabled it, but one of the main reasons is because it's not stable! If we make it stable there will surely a lot more websites enable it. The biggest problem with Memcached is it doesn't always clear caches of related SQL queries, so we don't see data on website somewhere up-to-date, this will cause a lot of further problems with the functioning of the website.

For now once enabled Memcached merchants face lot of problems, you only see memcached enabled on small websites because they are almost static websites (less orders, less customers...) so these merchants are not aware of problems with memcached. But for large websites (where memecached is really necessary) they quickly see problems with memcached and disable it!

**Comment by matteolavaggi** [comment]
@kpodemski i've made some dee test using all type of caching system, and yes, today prestashop does not support memcached at all. BTW this is really a great pity, memcached dedicated server help to reduce of about 60% loading time also under full page cache (also memcached) plugin. 
The only thing I can recommend (I don't think the development team will ever fix this problem because it requires major overhaul work and it is evident that they are not working in this direction) is to keep all prestasshop caches disabled, even the smarty and use a full page cache module + a cdn for static assets. Even CCC optimizations in my opinion are not a good solution, as PS is written huge js and css files are generated complete with all modules installing everything. If you use devices with medium-fast connections with http2 push it is better to load multiple files in parallel than a single giant. With these precautions it is possible to have page loads of less than 300ms, to go further and download the server you need a different code or a different platform.

**Comment by ets-soft** [comment]
@matteolavaggi  I can confirm the problem still here in PrestaShop 1.7.8.1
@khouloudbelguith won't PrestaShop manage to fix this? this is really big issue!

**Comment by matteolavaggi** [comment]
This will not be fixed ... This is a function that was never developed,
forgot it has no sense to hope in something that for the last 4 years was
not released. Don't enable, maybe some dev will remove from BO ...

Il Mar 30 Nov 2021, 11:09 ets-soft ***@***.***> ha scritto:

**Comment by kpodemski** [comment]
@ets-soft 

I have a working fix for this with simple `Cache::flush()` when you save the block, but the issue with the Memcached in PrestaShop is a lot bigger. We need to re-think the strategy for the Memcached and how to implement it within a module. For example, in `ps_linklist` we probably should be able to run something like: `Cache::flush('ps_linklist_*')`.

**Comment by ets-soft** [comment]
> @ets-soft
> 
> I have a working fix for this with simple `Cache::flush()` when you save the block, but the issue with the Memcached in PrestaShop is a lot bigger. We need to re-think the strategy for the Memcached and how to implement it within a module. For example, in `ps_linklist` we probably should be able to run something like: `Cache::flush('ps_linklist_*')`.

@kpodemski  ps_linklist is just an example. This problem frequently happens for other data entities, you save in BO but the update doesn't reflect on FO.

@matteolavaggi I found that PrestaShop detects cached data tables using regular expression that contain a "`" character in the regular expression before and after table names of **SQL Select queries**. So if Select query doesn't have "`" character it won't be deleted when the **Update queries** are executed that is one of things causing the problem! See classes/cache/Cache.php

`public function getTables($string)
    {
        if (preg_match_all('/(?:from|join|update|into)\s+`?(' . _DB_PREFIX_ .
            '[0-9a-z_-]+)(?:`?\s{0,},\s{0,}`?(' . _DB_PREFIX_ .
            '[0-9a-z_-]+)`?)?(?:`|\s+|\Z)(?!\s*,)/Umsi', $string, $res)) {
            foreach ($res[2] as $table) {
                if ($table != '') {
                    $res[1][] = $table;
                }
            }

            return array_unique($res[1]);
        } else {
            return false;
        }
    }`


So to fix the problem you need to make sure all Select queries are rounded by "`" character. You can use an IDE software such as PHPStorm to found and replace all NON `"`" `table names in Select queries with "`" table names in all PrestaShop source code using IDE's regular expression replacement feature. 

Here is my search expression for PHPStorm:

Find: \s(\'\s*?\.\s*?_DB_PREFIX_\s*?\.\s*\'[a-zA-Z0-9_]+?|\"\s*?\.\s*?_DB_PREFIX_\s*?\.\s*?\"[a-zA-Z0-9_]+?)\s
Replace with: " $1 " (make sure you have space before and after $1)

Thousands queries will be replaced! You can try this with copy of your website to make sure everything works after replacement.

Once you're done with the replacement you will see the memcached works properly in most cases but lot all! I have not figured out the other problems causing caches to not be deleted! But with this PrestaShop is usable with memcached (sometime if you see caches are not deleted, you must manually flush the memcached server)


*Suggestion: You can try to modify the getTables function to capture both non "`" table names and table names with "`" so you don't have to do the replacements.

Good luck!

**Comment by kpodemski** [comment]
Thank you for your feedback @ets-soft

**Comment by matteolavaggi** [comment]
Memcached cluster server  manage also full page cache and lot more feature,
so clean it is not a solution. This workaround can be used in dev server
but for product nobody can rely on a "maybe can broke your store"
configuration. What I can see here is a very early alpha function (the
cache) released like stable in 5 years code. This is absolutely a
nightmare.

Il Mar 30 Nov 2021, 11:46 ets-soft ***@***.***> ha scritto:

**Comment by mohaidoss** [comment]
Hi I just wanted to let you know that the same happened for me with the version 8.2 using APC for caching. 
Once I turned off the cache, it appears that my former edits were indeed taken into account.

So funnily, I just turned the cache off and on

I think that this bug should be mentioned in the BO as note in the link list page, or since it may affect other entities it's better to mention it in the Performance->Cache section

Edit: Turning off/on wasn't enough, I had to clear off the cache with `apcu_clear_cache()` before enabling caching.

Edit: Seems like it affects [cart entity](https://www.prestashop.com/forums/topic/549728-memecacheapc-caching-issue-customers-checkout-details-not-passed-through-to-backoffice/)

**Comment by boo-code** [comment]
Reproduces on 9.2, and the cause is the one @ets-soft named in the first post - writes that go through
Doctrine do not clear the legacy SQL cache. That hypothesis was never followed up, so here it is
measured.

The link widget shows it because the two halves use different connections. The front office reads
through `LegacyLinkBlockRepository::getByIdHook()`, which calls `Db::executeS()` and is therefore
cached. The back office writes through `LinkBlockRepository`, which uses a Doctrine `QueryBuilder` and
never touches `Db`. `Db::execute()` and `Db::delete()` are what call `Cache::deleteQuery()`, so the
write leaves the cached read exactly where it was.

Caching that read and then performing the same `UPDATE` four ways:

```
                                                              before        after
Db::execute (the contract that already works)                 invalidated   invalidated
DBAL executeStatement, no parameters       -> exec()          STALE         invalidated
DBAL executeStatement with parameters      -> prepare()       STALE         invalidated
DBAL QueryBuilder update (what the module does)               STALE         invalidated
Doctrine ORM EntityManager::flush()                           STALE         invalidated
a SELECT through Doctrine                                     untouched     untouched
```

Each row checks the entry is cached immediately before the write, so STALE means it really survived.
The `Db` row is the control: the identical statement on the other connection has always invalidated,
which makes this a contract that one connection honours and the other does not, rather than a missing
feature. Disabling the fix and rerunning puts the three Doctrine rows back to STALE and leaves the
`Db` row unchanged.

That also explains why this is not confined to the link widget, as @ets-soft said and as
@mohaidoss saw on carts under APC in 8.2: it applies to every entity written through Doctrine and read
through the legacy `Db`, which is a growing set as controllers are migrated.

**One correction, because people are still acting on it.** The advice further up the thread is that
`Cache::getTables()` only matches table names written between backticks, and that the cure is a mass
find and replace across thousands of queries. That is not the cause. The opening backtick in that
pattern is optional and the closing alternation accepts whitespace or end of string, so both
spellings resolve the same:

```
FROM ps_link_block lb INNER JOIN ps_link_block_shop lbs   -> ps_link_block, ps_link_block_shop
FROM `ps_link_block` lb INNER JOIN `ps_link_block_shop`   -> ps_link_block, ps_link_block_shop
UPDATE ps_link_block SET id_hook = 2                      -> ps_link_block
UPDATE `ps_link_block` SET `id_hook` = 2                  -> ps_link_block
SELECT * FROM ps_link_block                               -> ps_link_block
```

The first line is the real unbackticked query from `LegacyLinkBlockRepository`, and it is registered
correctly. The regex is byte identical in 1.7.6.5 and in 9.2.x, so it was not the cause when that
advice was written either.

A branch is ready. It follows `SetSessionTimeZoneMiddleware`, which already mirrors a legacy `Db`
behaviour on the Symfony connection as a `doctrine.middleware`, and covers both ways a statement
reaches the driver - `exec()` without parameters, `prepare()` plus `Statement::execute()` with them,
which is the path the QueryBuilder takes, so covering only `exec()` would miss the reported case.
Reads are handed back undecorated, and a shop with the legacy cache off, which is the default, stops
at a boolean and never builds a cache backend; with it on the added work per statement is one anchored
`preg_match`, 53 ns measured.

Two notes on scope. This is the write path never reaching the cache. #25185 is a different defect
inside the cache - the table-to-query map is maintained by read-modify-write, so the second query a
request caches on a table drops another request's entries - and it is not addressed here; both are
real and they compound. And #42530 fixes one instance of today's problem by hand in
`ProductRepository`; this covers that call site along with the rest.
